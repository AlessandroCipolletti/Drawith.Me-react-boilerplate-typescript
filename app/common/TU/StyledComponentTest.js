import React from 'react'
import { render } from 'react-testing-library'

// Import de Theme Provider
import { ThemeProvider } from 'styled-components'
import Theme from 'common/Theme'

const StyledComponentTest = (Component, componentName, componentTag) => {
  if (!Component || !componentName || !componentTag) {
    return
  }

  describe(`<${componentName} />`, () => {
    const renderComponent = (props = {}) =>
      render(
        <ThemeProvider theme={Theme}>
          <Component {...props}>
            Content
          </Component>
        </ThemeProvider>,
      )

    it(`should render an <${componentTag}> tag`, () => {
      const { container } = renderComponent()
      expect(container.querySelector(componentTag)).not.toBeNull()
    })

    it('should have a class attribute', () => {
      const { container } = renderComponent()
      expect(container.querySelector(componentTag).hasAttribute('class')).toBe(
        true,
      )
    })

    it('should adopt a valid attribute', () => {
      const id = 'test'
      const { container } = renderComponent({ id })
      expect(container.querySelector(componentTag).id).toEqual(id)
    })

    it('should not adopt an invalid attribute', () => {
      const { container } = renderComponent({ attribute: 'test' })
      expect(container.querySelector(`${componentTag}[attribute="test"]`)).toBeNull()
    })
  })
}

export default StyledComponentTest
