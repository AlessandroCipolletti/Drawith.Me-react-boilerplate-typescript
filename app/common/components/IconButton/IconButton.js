import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Theme from 'common/Theme'

import Icon from 'common/components/Icon'

const IconButton = ({
  onClick = () => {},
  disabled = false,
  color = Theme.palette.gray[9],
  float = 'none',
  icon,
}) => (
  <Wrapper
    float={float}
    disabled={disabled}
    onClick={!disabled ? onClick : () => {}}
    data-testid="IconButton"
  >
    <Icon
      name={icon}
      color={color}
      size={Theme.spacing.size6}
    />
  </Wrapper>
)

IconButton.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  color: PropTypes.string,
  float: PropTypes.oneOf(['right', 'left']),
  icon: PropTypes.string.isRequired,
}

export default IconButton

export const Wrapper = styled.div`
  width: ${({ theme }) => theme.spacing.size7};
  height: ${({ theme }) => theme.spacing.size7};
  float: ${props => props.float};
  box-sizing: border-box;
  padding: 6px;
  cursor: ${props => props.disabled ? 'auto' : 'pointer' };
  opacity: ${props => props.disabled ? 0.6 : 1 };
`
