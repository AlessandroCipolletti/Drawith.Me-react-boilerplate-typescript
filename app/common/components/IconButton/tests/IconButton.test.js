import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import StyledComponentTest from 'common/TU/StyledComponentTest'

// Import de Theme Provider
import { ThemeProvider } from 'styled-components'
import Theme from 'common/Theme'

import IconButton, { Wrapper } from '../IconButton'

StyledComponentTest(Wrapper, 'Wrapper', 'div')

describe('<IconButton Wrapper />', () => {
  const renderComponent = (props = {}) => (
    render(
      <ThemeProvider theme={Theme}>
        <Wrapper {...props} />
      </ThemeProvider>,
    )
  )

  it('Should get right cursor css depending on "disabled" prop = false', () => {
    const disabled = false

    const {
      container: { firstChild },
    } = renderComponent({ disabled })
    const style = window.getComputedStyle(firstChild)

    expect(style.cursor).toBe('pointer')
  })

  it('Should get right cursor css depending on "disabled" prop = true', () => {
    const disabled = true

    const {
      container: { firstChild },
    } = renderComponent({ disabled })
    const style = window.getComputedStyle(firstChild)

    expect(style.cursor).toBe('auto')
  })

  it('Should get right opacity css depending on "disabled" prop = false', () => {
    const disabled = false

    const {
      container: { firstChild },
    } = renderComponent({ disabled })
    const style = window.getComputedStyle(firstChild)

    expect(parseFloat(style.opacity)).toBe(1)
  })

  it('Should get right opacity css depending on "disabled" prop = true', () => {
    const disabled = true

    const {
      container: { firstChild },
    } = renderComponent({ disabled })
    const style = window.getComputedStyle(firstChild)

    expect(parseFloat(style.opacity)).toBeLessThan(1)
  })
})

describe('<IconButton />', () => {

  const renderComponent = (props = {}) => (
    render(
      <ThemeProvider theme={Theme}>
        <IconButton {...props} />
      </ThemeProvider>,
    )
  )

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')
    const icon = 'icom'

    renderComponent({ icon })

    expect(spy).not.toHaveBeenCalled()
  })

  it('Expect log an error in console if called without icon prop', () => {
    const spy = jest.spyOn(global.console, 'error')

    renderComponent()

    expect(spy).toHaveBeenCalledTimes(2)
  })

  it('Should render and match the snapshot with default props', () => {
    const onClick = () => {}

    const {
      container: { firstChild },
    } = renderComponent({ onClick })

    expect(firstChild).toMatchSnapshot()
  })

  it('Should have the right onClick value depending on "disabled" prop = false', () => {
    const disabled = false
    const onClick = jest.fn()

    const {
      container: { firstChild },
    } = renderComponent({ disabled, onClick })
    fireEvent.click(firstChild)

    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('Should have the right onClick value depending on "disabled" prop = true', () => {
    const disabled = true
    const onClick = jest.fn()

    const {
      container: { firstChild },
    } = renderComponent({ disabled, onClick })
    fireEvent.click(firstChild)

    expect(onClick).toHaveBeenCalledTimes(0)
  })
})
