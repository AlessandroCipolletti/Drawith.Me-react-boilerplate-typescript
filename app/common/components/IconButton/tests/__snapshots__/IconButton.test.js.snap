// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> Should render and match the snapshot with default props 1`] = `
<div
  class="IconButton__Wrapper-sc-15xi3uz-0 ixFaiP"
  data-testid="IconButton"
>
  <div
    class="Icon__Wrapper-xawg7a-0 hKOtws"
    color="#f8f9fa"
    role="button"
    size="32"
    tabindex="0"
  />
</div>
`;
